services/relayer/internal/msgobserver/eth/observer/observer_impl_v1.go

```diff
@@ -145,25 +146,33 @@ func (o *ObserverImplV1) StopObserve() error {
 
 // routine is the observe routine.
 func (o *ObserverImplV1) routine(chainID *big.Int, chainAP string, addr common.Address, end chan bool) {
-	chain, err := ethclient.Dial(chainAP)
-	if err != nil {
-		logging.Error(err.Error())
-		return
-	}
-	defer chain.Close()
+	for {
+		chain, err := ethclient.Dial(chainAP)
+		if err != nil {
+			logging.Error(err.Error())
+			return
+		}
+		defer chain.Close()
 
-	sfc, err := functioncall.NewSfc(addr, chain)
-	if err != nil {
-		logging.Error(err.Error())
-		return
-	}
+		sfc, err := functioncall.NewSfc(addr, chain)
+		if err != nil {
+			logging.Error(err.Error())
+			return
+		}
 
-	observer, err := NewSFCBridgeObserver(chainID.String(), addr.String(), sfc, o.mq, end)
-	if err != nil {
-		logging.Error(err.Error())
-		return
+		observer, err := NewSFCBridgeObserver(chainID.String(), addr.String(), sfc, o.mq, end)
+		if err != nil {
+			logging.Error(err.Error())
+			return
+		}
```

Comment:
Is there a reason we would repeat these steps (lines 150-167) if the observer fails to start on line 168? Could the for loop start after line 167 instead?

Reply:
The idea here is that if a network loss appears then we need to basically dial again by closing the old connection, open a new one + refresh the sfc and observer with the new ethclient instance. If we just loop after line 167, my suspicion is that the observer will just reuse the old connection.

Reply:
Ah, I assumed the `observer.Start()` call here would execute on a separate goroutine, but instead, it is a blocking and long-running operation so resetting the connection makes sense to me now.